### PR TITLE
feat: add MaxTasks middleware

### PR DESCRIPTION
**Pull request by JonathanGriffe**

### CODE REVIEW

File: remoulade/middleware/max_tasks.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        self.logger = get_logger(__name__, type(self))
+        self.max_tasks = max_tasks
+        self.tasks_count = 0
+        self.lock = threading.Lock()
+
+    def after_worker_thread_process_message(self, broker, thread):
+        with self.lock:
+            self.tasks_count += 1
+            if self.tasks_count == self.max_tasks:
+                self.logger.warning(
```
Comment (antoinerabany) [inline remoulade/middleware/max_tasks.py:26]: I think it's should just be info, it's expected.

File: tests/middleware/test_max_tasks.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        nonlocal result
+        result += 1
+        time.sleep(1)
+
+    stub_broker.declare_actor(do_work)
+
+    for _ in range(300):
+        do_work.send()
+
+    main()
```
Comment (antoinerabany) [inline tests/middleware/test_max_tasks.py:51]: check the return code also

File: tests/middleware/test_max_tasks.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    def do_work():
+        nonlocal result
+        result += 1
+
+    stub_broker.declare_actor(do_work)
+
+    for _ in range(30):
+        do_work.send()
+
+    time.sleep(0.1)
```
Comment (antoinerabany) [inline tests/middleware/test_max_tasks.py:27]: Is it necessary to do a sleep, we cannot use the woker special fucntion to wait ?

**Review by antoinerabany — CHANGES_REQUESTED** [review]

File: tests/middleware/test_max_tasks.py (hunk 3 above)
Comment (JonathanGriffe) [reply tests/middleware/test_max_tasks.py:27]: The problem here is that only the WorkerThread is stopped and not the Worker. The loop that stops the Worker when one of its WorkerThread is stopped is inside the cli and therefore not executed. As a result, worker.join will never return and cannot be used here. Neither can broker.join as some messages will be left waiting.

**Pull request merged**